### DISCUSSION TITLE
Including x86 and x64 assemblies in build

### DISCUSSION THREAD

**Discussion by TonyValenti** [cefsharp/CefSharp#3617] [Q&A]
I have a .NET 4.x app that I'm trying to upgrade to the latest version of CEFSharp but I'm having troubles.
I was previously using v79 and was including the x64 and x86 assemblies in my app and then dynamically picking the right DLLs to load at runtime.

Now, after upgrading my packages, I cannot get a successful build.

First, I get this error:
```
CefSharp.Common is unable to proceeed as your current PlatformTarget is 'AnyCPU'. To target AnyCPU please read https://github.com/cefsharp/CefSharp/issues/1714. Alternatively change your PlatformTarget to x86 or x64 and the relevant files will be copied automatically.
```

So, per https://github.com/cefsharp/CefSharp/issues/1714, I add :
```
  <PropertyGroup>
    <CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>
  </PropertyGroup>
```

Then I get this error:
```
CefSharp copied resources/libs for AnyCPU where the resulting exe was , this is a known issue with SDK Style projects see https://github.com/dotnet/sdk/issues/1560 please explicitly set PlatformTarget in your proj file.
```

I'm going in circles and not making progress.

I'm a one-man company but I'm willing to pay @amaitland or any other person a few hundred dollars to help me make my app buildable with x86 and x64 binaries.  If you're interested, book a time with me:
https://www.mattermovers.net/bookings-checkout/general-meeting/book

or call me.  402-509-9270.

**Comment by amaitland** [comment]
Add `<PlatformTarget>AnyCPU</PlatformTarget>` to your project file should fix the problem. 

If setting PlatformTarget doesn't resolve the problem then set `<CefSharpPlatformTargetOverride>AnyCPU</CefSharpPlatformTargetOverride>` which if set will disable the build time checks. 

https://github.com/cefsharp/CefSharp/blob/cefsharp/90/NuGet/CefSharp.Common.targets#L126
https://github.com/cefsharp/CefSharp/blob/cefsharp/90/NuGet/CefSharp.Common.targets#L389

To get a slightly better understanding of what's going on if you call CefSharpAfterBuildDiagnostic target after build it'll output some diagnostic information.

https://github.com/cefsharp/CefSharp/blob/cefsharp/90/NuGet/CefSharp.Common.targets#L53

> CefSharp copied resources/libs for AnyCPU where the resulting exe was , this is a known issue with SDK Style projects see https://github.com/dotnet/sdk/issues/1560 please explicitly set PlatformTarget in your proj file.

I'm curious as to how PlatformTarget was empty, are you able to provide a copy of your project file? (Minus any sensitive information).

**Reply by TonyValenti** [reply]
@amaitland - Unfortunately that isn't working.  And, oddly, now I'm getting CEFSharp errors from projects that don't reference CEFSharp.  For example:

Here's the error:
```
27>------ Build started: Project: AlphaDrive.Authentication(Windows), Configuration: Debug Any CPU ------
27>C:\Users\Tony Valenti\.nuget\packages\cefsharp.common\90.6.70\buildTransitive\CefSharp.Common.targets(363,5): error : CefSharp.Common is unable to proceeed as your current PlatformTarget is 'AnyCPU'. To target AnyCPU please read https://github.com/cefsharp/CefSharp/issues/1714. Alternatively change your PlatformTarget to x86 or x64 and the relevant files will be copied automatically.
27>Done building project "AlphaDrive.Authentication(Windows).csproj" -- FAILED.
```
And here's the csproj:
```
<Project Sdk="Microsoft.NET.Sdk.WindowsDesktop">
  <PropertyGroup>
    <UseWPF>true</UseWPF>
    <RootNamespace>AlphaDrive.Authentication</RootNamespace>
  </PropertyGroup>
  <Import Project="..\AlphaDrive.Build.DeployComponent\AlphaDrive.Build.DeployComponent.targets" Label="Shared" />
  <ItemGroup>
    <Reference Include="Telerik.Windows.Controls">
      <HintPath>..\ASSEMBLIES\TELERIK\WPF45\Telerik.Windows.Controls.dll</HintPath>
    </Reference>
    <Reference Include="Telerik.Windows.Controls.Input">
      <HintPath>..\ASSEMBLIES\TELERIK\WPF45\Telerik.Windows.Controls.Input.dll</HintPath>
    </Reference>
    <Reference Include="Telerik.Windows.Controls.Navigation">
      <HintPath>..\ASSEMBLIES\TELERIK\WPF45\Telerik.Windows.Controls.Navigation.dll</HintPath>
    </Reference>
    <Reference Include="Telerik.Windows.Controls.RibbonView">
      <HintPath>..\ASSEMBLIES\TELERIK\WPF45\Telerik.Windows.Controls.RibbonView.dll</HintPath>
    </Reference>
    <Reference Include="Telerik.Windows.Data">
      <HintPath>..\ASSEMBLIES\TELERIK\WPF45\Telerik.Windows.Data.dll</HintPath>
    </Reference>
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\..\MC.Clio.Net4x\MC.Clio.Caching.EFCore\MC.Clio.Caching.EFCore.csproj" />
    <ProjectReference Include="..\AlphaDrive.Authentication\AlphaDrive.Authentication.csproj" />
    <ProjectReference Include="..\AlphaDrive.Core\AlphaDrive.Core.csproj" />
    <ProjectReference Include="..\AlphaDrive.Core.Windows\AlphaDrive.Core.Windows.csproj" />
    <ProjectReference Include="..\..\MC.Clio.Net4x\MC.Clio.Authentication.Wpf.CefSharp\MC.Clio.Authentication.Wpf.CefSharp.csproj" />
    <ProjectReference Include="..\..\MC.Clio.Net4x\MC.Clio.NET461\MC.Clio.csproj" />
    <ProjectReference Include="..\..\MC.Clio.Net4x\MC.Clio.Authentication.Wpf\MC.Clio.Authentication.Wpf.csproj" />
    <ProjectReference Include="..\AlphaDrive.Data.Connectivity\AlphaDrive.Data.Connectivity.csproj" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="PostSharp" Version="6.6.16" />
  </ItemGroup>
</Project>
```

Can i please pay you to jump on a screen share with me and solve this issue?  You definitely have a lot more experience with this stuff than I do.

**Reply by TonyValenti** [reply]
Also, here's the DeployComponent.targets file if you're interested:
```
<?xml version="1.0" encoding="utf-8"?>
<Project xmlns="http://schemas.microsoft.com/developer/msbuild/2003" ToolsVersion="4.0" InitialTargets="All">

  <ItemGroup>
    <Compile Include="..\AlphaDrive.Build.DeployComponent\__InternalAssemblyInfo.g.cs">
      <Visible>False</Visible>
    </Compile>

    <Compile Include="..\AlphaDrive.Build.DeployComponent\AssemblyInfo.cs">
      <Visible>False</Visible>
    </Compile>
    
  </ItemGroup>

  <PropertyGroup>
    <TargetFramework>net472</TargetFramework>
    <GenerateAssemblyInfo>false</GenerateAssemblyInfo>
    <DebugType>full</DebugType>
    <OutputPath>bin\</OutputPath>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
  
      
    

  
  
    <Target Name="All">
        <PropertyGroup>
            <!--
            FROM https://github.com/Microsoft/msbuild/issues/2468
            -->
            <PowerShellExe Condition="'$(OS)' == 'Windows_NT' AND '$(PowerShellExe)'=='' ">%WINDIR%\System32\WindowsPowerShell\v1.0\powershell.exe</PowerShellExe>
            <PowerShellExe Condition="'$(OS)' == 'Unix' AND '$(PowerShellExe)'=='' ">/usr/local/bin/pwsh</PowerShellExe>
            
            <ScriptLocation Condition=" '$(ScriptLocation)'=='' ">$(SolutionDir)AlphaDrive.Build.DeployComponent\Deploy.PS1</ScriptLocation>
            <ScriptMethod Condition=" '$(ScriptMethod)'=='' ">Deploy</ScriptMethod>
      
            <DeploySource Condition=" '$(DeploySource)'=='' ">$(TargetDir)</DeploySource>
      
            
            <DeployDest Condition="'$(OS)' == 'Windows_NT' AND '$(DeployDest)'=='' ">$(SolutionDir)\__BUILD\Current</DeployDest>
            <DeployDest Condition="'$(OS)' == 'Unix' AND '$(DeployDest)'=='' ">$(SolutionDir)\AlphaDrive.Mac\bin\AlphaDrive.app\Contents\MonoBundle</DeployDest>
      
            <!--
                <DeployDest Condition=" '$(DeployDest)'=='' ">\\Dev3-pc\c$\Users\Dev3\AppData\Local\Clio\app-18.12.12</DeployDest>
            -->
        </PropertyGroup>

    </Target>

  <PropertyGroup>
    <!--
    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
    <OutDir>$(DeployDest)</OutDir>
    -->
  </PropertyGroup>

    <Target Name="__COPY_TO_BUILD" AfterTargets="Build">
   
        <Message Text="Coping to __BUILD folder..."/>

    
        <Message Text="&quot;$(PowerShellExe)&quot; -NonInteractive -executionpolicy Unrestricted -command &quot;&amp; { Import-Module &#39;$(ScriptLocation)&#39;; $(ScriptMethod) &#39;$(DeploySource)&#39; &#39;$(DeployDest)&#39;} &quot;"/>
    
        <Exec Command="&quot;$(PowerShellExe)&quot; -NonInteractive -executionpolicy Unrestricted -command &quot;&amp; { Import-Module &#39;$(ScriptLocation)&#39;; $(ScriptMethod) &#39;$(DeploySource)&#39; &#39;$(DeployDest)&#39;} &quot;" />

  </Target>

  <Target Name="__Clean" AfterTargets="Clean">
    <Delete Files="$(DeploySource)" />
    <RemoveDir Directories="$(DeploySource)" />
    <RemoveDir Directories="$(DeploySource)..\obj\" />
    <RemoveDir Directories="$(ProjectDir)$(BaseIntermediateOutputPath)" /> <!-- obj -->
  </Target>
  
</Project>

```

**Reply by amaitland** [reply]
> And, oddly, now I'm getting CEFSharp errors from projects that don't reference CEFSharp. For example:

The Nuget packages now support buildTransative which is enabled by default. The easiest option is to add a Directory.Build.props file. https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#directorybuildprops-example

See https://github.com/cefsharp/CefSharp/issues/3319#issuecomment-818354408 for additional information. 

An example `Directory.Build.props` for your use case would look something like

```xml
<Project>
 <PropertyGroup>
   <CefSharpAnyCpuSupport>true</CefSharpAnyCpuSupport>
  <CefSharpPlatformTargetOverride>AnyCPU</CefSharpPlatformTargetOverride>
 </PropertyGroup>
</Project>
```

Where you put the file matters, see https://docs.microsoft.com/en-us/visualstudio/msbuild/customize-your-build?view=vs-2019#search-scope

> Can i please pay you to jump on a screen share with me and solve this issue? You definitely have a lot more experience with this stuff than I do.

Try adding a `Directory.Build.props` file first, if that doesn't work then we can organise a time to sort this out.

**Reply by TonyValenti** [reply]
@amaitland Thanks!  That works!
Let me send you some paypal love.  What currency do you prefer? USD or AUD?

**Reply by amaitland** [reply]
> Let me send you some paypal love. What currency do you prefer? USD or AUD?

Thank you! `AUD` is preferable, though honestly either is fine. https://paypal.me/AlexMaitland

I've been thinking about how to improve the `CefSharpAnyCpuSupport` use case for `Library` projects that support `buildTransitive`, I've created #3622 to track improving that use case. I don't anticipate any breaking changes, the `Directory.Build.props` example provided above should work in all cases.
